web/src/stream_popover.js

```diff
@@ -214,7 +215,77 @@ function build_stream_popover(opts) {
     });
 }
 
-export function build_move_topic_to_stream_popover(
+async function get_message_placement_from_server(
+    current_stream_id,
+    topic_name,
+    current_message_id,
+) {
+    return new Promise((resolve) => {
+        message_edit.is_message_oldest_or_newest(
+            current_stream_id,
+            topic_name,
+            current_message_id,
+            (is_oldest, is_newest) => {
+                if (is_newest) {
+                    resolve("last");
+                } else if (is_oldest) {
+                    resolve("first");
+                } else {
+                    resolve("intermediate");
+                }
+            },
+        );
+    });
+}
+
+async function get_message_placement_in_conversation(
```

Comment:
So here is the offline-first approach I have tried to implement by following your example.